canonical/vanilla-framework
# Migrate Percy from CircleCI to Github Actions

## Done

Switches Percy integration from CircleCI to Github Actions

Closes [WD-11242](https://warthogs.atlassian.net/browse/WD-11242)

### The problem
We are currently using CircleCI to execute Percy visual testing, because executing `percy snapshot` requires an environment variable `PERCY_TOKEN` to be set. However, repo secrets are not exposed to fork contributors for security reasons. CircleCI is used as a workaround for this, as it knows the Percy secret and is triggered on PR. 

However, this disjoint approach causes issues with linking Percy build state to the PR. The PR CircleCI check success only means that the snapshots were uploaded to Percy; it does not necessarily mean that Percy diffs were approved. 

### The solution
This PR is the first step towards solving the above problem. This migrates us to using GHA to start Percy, allowing us to report Percy diff state in the PR at a later date and also run Percy more selectively (such as only when certain files change) at a later date.

#### The approach
Two workflows are used: `pr-percy-copy-diff-files`, and `pr-percy-create-screenshots`. This follows [this approach](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/) by GH security team.

##### Copy diff files
Runs in an unprivileged context on `pull_request` to `main`. It checks out the `templates/docs/examples/` and `scss/` directories, notes the PR # & commit signature (as suggested by the GH security article above), and uploads it as an artifact to Github Actions associated with the workflow run.

##### Create screenshots
Runs in a privileged context on `workflow_run` trigger. Runs after completion of the copy diff files workflow. It downloads the artifacts from the previous workflow and replaces the matching files in `main` with the files from the artifacts. It then starts a local Flask server and uploads screenshots to Percy.

Note that for this workflow to run, `pr-percy-create-screenshots.yml` must be in the default branch of the target repo (`canonical/vanilla-framework`) at the time of the first workflow's completion. Thus, **it will not run on this repo until this PR is merged** and a second PR is raised to test its functionality.

## QA
As mentioned above, screenshots will not be created by this PR as `pr-percy-create-snapshots.yml` is not yet in our main branch. However, the PR workflow will run and upload diff artifacts based on this PR commit. The following should occur after this PR is raised:

- Copy diff files workflow is started & completed
- After completion of the diff files workflow, an artifact is available for download in the workflow summary page with the name `web-artifact-{workflow_run_id}`.  Opening this artifact will reveal a `build.tar.gz` file containing:
    - examples/ (the `templates/docs/examples/` folder from the PR code)
    - scss/ (the `scss/` folder from the PR code)
    - pr_head_sha.txt (contains the signature of the last commit in the PR code
    - pr_num.txt (contains the number of this PR, i.e., `https://github.com/canonical/vanila-framework/pull/{pr_number}`)

To fully test this workflow in a fork PR context, a [separate repository](https://github.com/canonical/test-web-percy-gha-migration) was created. An example workflow is shown:

1. [PR](https://github.com/canonical/test-web-percy-gha-migration/pull/12) created
2. Copy diff files workflow [runs](https://github.com/canonical/test-web-percy-gha-migration/actions/runs/9112576128) & creates [artifact](https://github.com/canonical/test-web-percy-gha-migration/actions/runs/9112576128/artifacts/1509357373)
3. Screenshot creation workflow [runs](https://github.com/canonical/test-web-percy-gha-migration/actions/runs/9112579120)
4. Status check [percy_upload](https://github.com/canonical/test-web-percy-gha-migration/runs/25052344643) is applied to the [PR](https://github.com/canonical/test-web-percy-gha-migration/pull/12), including a link to the [Percy build](https://percy.io/bb49709b/test-vanilla-gha-migration/builds/34252438/changed/1874590093?browser=chrome&browser_ids=56&group_snapshots_by=similar_diff&subcategories=unreviewed%2Cchanges_requested&test_case_ids=&viewLayout=side-by-side&viewMode=new&width=375&widths=375)

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
  - if CSS class names are not changed it can be bugfix relesase (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)

### Screenshots
Upload & Snapshot status checks on PR:
![image](https://github.com/canonical/vanilla-framework/assets/46915153/0b85c034-4f4f-4c09-b951-7ca36ef7b0e1)
Percy diff:
![image](https://github.com/canonical/vanilla-framework/assets/46915153/21c23faf-7e82-4cdd-b99f-96c906988157)
Upload completed status check on PR with Percy build details:
![image](https://github.com/canonical/vanilla-framework/assets/46915153/5ced31ea-39e0-480b-b28f-4c5600af2d26)



[WD-11242]: https://warthogs.atlassian.net/browse/WD-11242?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
